### ISSUE TITLE
Runtime configurable logger

### ISSUE THREAD

**Issue by sethvargo** [abcxyz/pkg#193]
### TL;DR

**Philosophical question:** Should we allow end-users to control log level and log format via CLI flags?

The current CLI architecture does not provide a great mechanism for setting log levels<sup>1</sup>, and slog does not provide an easy way to change a formatter (handler) after creation<sup>2</sup>.

I've generally preferred loggers to be configured _as_ the application boots, because otherwise you risk not having a configured logger earlier in the stack. There were [a few requests ](https://github.com/abcxyz/pkg/issues/121)to configure the logger via CLI flags (e.g. `-log-level`), but the current architecture doesn't actually support that.

### Log Levels

<sup>1</sup> **Log Levels** - We had to build a [custom handler](https://github.com/abcxyz/pkg/blob/9b36be5112d6211f33d81a8ebcb29693e66a3809/logging/handler.go#L24) that wraps the provided handler to allow for runtime changes of the level. Even then, it's not as simple as changing the level, since `slog/logger.Handler()` returns an interface. So we [have a custom function that does a typecast](https://github.com/abcxyz/pkg/blob/9b36be5112d6211f33d81a8ebcb29693e66a3809/logging/logger.go#L131), resulting in usage like:

```go
logger := logging.New(...)
logging.SetLevel(logger, slog.LevelDebug)
```

This is a bit weird because it (safely) modifies the logger in place, but that might be unexpected for some people. Despite this, we [added a CLI flag for configuring the log level](https://github.com/abcxyz/pkg/blob/9b36be5112d6211f33d81a8ebcb29693e66a3809/cli/flags.go#L878), which accepts a logger and configures the log-level with the given level:

```go
func Flags() *cli.FlagSet {
  //

  f.LogLevelVar(&cli.LogLevelVar{
    Logger: logger,
  })
}
```

However, largely because of my stupidity, we have no way to _get_ a logger inside the `Flags` package. To accomplish that, we'd need to change the type signature of `Flags` to accept a context:

```diff
type Command interface {
-  Flags() *FlagSet
+  Flags(context.Context) *FlagSet
}
```

And then pull the logger out of the context. That's probably a good change to make anyway, but it feels disruptive. We could do it in a backwards-compatible way by moving `Flags()` into its own interfaces and then doing type assertions later down the stack:

```go
type CommandFlags interface {
  Flags() *FlagSet
}

type CommandFlagsContext interface {
  Flags(context.Context) *FlagSet
}
```

```go
func[T CommandFlags | CommandFlagsAny] getFlags(ctx context.Context, fn T) *FlagSet {
  if typ, ok := fn.(CommandFlagsContext); ok {
    return typ(ctx)
  }
  return fn(context.Background())
}
```

That would theoretically maintain backwards compat (although I haven't actually tested it). I'd rather just break compat, since I think we'll need context's in `Flags()` for other things in the future.

### Handlers (format)

<sup>2</sup> **Handlers** - In slog, formatting (json, text, yaml) is delegated to a [handler](https://pkg.go.dev/log/slog#Handler), which is a Go interface, and there's no clean way to change the handler a `slog.Logger` uses after instantiation.

We'd need to create a custom handler that embedded all expected format handlers inside. To change the format, we'd have to update our handler to point to the correct handler that maps to that format. It's technically possible, but it'd be extremely fragile and not flexible.

```go
type FormatHandler struct {
  activeKey string
  handlers map[string]slog.Handler
}

func NewFormatHandler(m map[string]slog.Handler, active string) *FormatHandler {
  return &FormatHandler{handlers: m, activeKey: active}
}

func (h *FormatHandler) Enabled(ctx context.Context, level slog.Level) bool {
  return h[activeKey].Enabled() // plus locking
}

func (h *FormatHandler) SetFormat(format string) {
  h.activeKey = format // plus validation, plus locking
}
```

and then every instantiation of that handler would need to create a text and json handler (plus any other custom formatters we'd want in the future):

```go
opts := &slog.HandlerOpts{}

slog.New(level, NewFormatHandler(map[string]slog.Handler{
  "json": slog.NewJSONHandler(w, opts),
  "text": slog.NewTextHandler(w, opts),
}, "text"))
```

## Summary

All-in, it's definitely possible to make this work. However, I'm starting to question the overall value in runtime-configurable loggers versus configuration from the environment variables + documentation. It's certainly _nice_ to be able to add `-v` or `-log-level=debug` to a CLI, but that's no _less_  convenient than `LOG_LEVEL=debug my-cli`.

**Comment by drevell** [comment]
environment vars seem acceptable to me too.
